File: apps/backend/src/routes/api/v1/notes.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';
import { noteService } from '../../../services/note.service';
import { authMiddleware } from '../../../middleware/auth';

const noteRouter = new Hono();

const createNoteSchema = z.object({
  title: z.string().min(1),
  content: z.string().optional(),
  visibility: z.enum(['private', 'public', 'shared']).optional(),
  tags: z.array(z.string()).optional(),
  folderId: z.string().nullable().optional(),
});

const updateNoteSchema = z.object({
  title: z.string().min(1).optional(),
  content: z.string().optional(),
  visibility: z.enum(['private', 'public', 'shared']).optional(),
  tags: z.array(z.string()).optional(),
  folderId: z.string().nullable().optional(),
});

noteRouter.get('/shared/:shareId', async (c) => {
  const shareId = c.req.param('shareId');
  const note = await noteService.getNoteByShareId(shareId);
  return c.json(note);
});

noteRouter.get('/', authMiddleware, async (c) => {
  const user = c.get('user');
  const archived = c.req.query('archived');
  const pinned = c.req.query('pinned');
  const folderId = c.req.query('folderId');
  const search = c.req.query('search');

  const filters: any = {};
  if (archived !== undefined) filters.archived = archived === 'true';
  if (pinned !== undefined) filters.pinned = pinned === 'true';
  if (folderId !== undefined) filters.folderId = folderId || null;
  if (search) filters.search = search;

  const notes = await noteService.getNotes(user.id, filters);
  return c.json(notes);
});

noteRouter.post('/', authMiddleware, zValidator('json', createNoteSchema), async (c) => {
  const user = c.get('user');
  const data = c.req.valid('json');

  const note = await noteService.createNote(user.id, data);
  return c.json(note);
});

noteRouter.get('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  const note = await noteService.getNoteById(id, user.id);
  return c.json(note);
});

noteRouter.put('/:id', authMiddleware, zValidator('json', updateNoteSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json');

  const note = await noteService.updateNote(id, user.id, data);
  return c.json(note);
});

noteRouter.delete('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  await noteService.deleteNote(id, user.id);
  return c.json({ message: 'Note deleted' });
});

noteRouter.put('/:id/archive', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const body = await c.req.json();
  const archived = body.archived ?? true;

  const note = await noteService.archiveNote(id, user.id, archived);
  return c.json(note);
});

noteRouter.put('/:id/pin', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const body = await c.req.json();
  const pinned = body.pinned ?? true;

  const note = await noteService.pinNote(id, user.id, pinned);
  return c.json(note);
});

noteRouter.post('/:id/share', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  const result = await noteService.shareNote(id, user.id);
  return c.json(result);
});

noteRouter.delete('/:id/share', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  await noteService.unshareNote(id, user.id);
  return c.json({ message: 'Note unshared' });
});

export default noteRouter;

File: apps/backend-python/src/routes/api/v1/notes.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from typing import Optional, List
from sqlalchemy.orm import Session
from src.config.database import get_db
from src.middleware.auth import get_current_user
from src.services.note_service import note_service
from src.db.models import User

router = APIRouter()


class CreateNoteSchema(BaseModel):
    title: str
    content: Optional[str] = ""
    visibility: Optional[str] = "private"
    tags: Optional[List[str]] = None
    folderId: Optional[str] = None


class UpdateNoteSchema(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    visibility: Optional[str] = None
    tags: Optional[List[str]] = None
    folderId: Optional[str] = None


class ArchiveSchema(BaseModel):
    archived: bool = True


class PinSchema(BaseModel):
    pinned: bool = True


@router.get("/shared/{share_id}")
def get_shared_note(share_id: str, db: Session = Depends(get_db)):
    return note_service.get_note_by_share_id(db, share_id)


@router.get("")
def list_notes(
    archived: Optional[str] = None,
    pinned: Optional[str] = None,
    folderId: Optional[str] = None,
    search: Optional[str] = None,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    filters = {}
    if archived is not None:
        filters["archived"] = archived == "true"
    if pinned is not None:
        filters["pinned"] = pinned == "true"
    if folderId is not None:
        filters["folderId"] = folderId or None
    if search:
        filters["search"] = search
    return note_service.get_notes(db, user.id, filters)


@router.post("")
def create_note(data: CreateNoteSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return note_service.create_note(db, user.id, data.model_dump())


@router.get("/{note_id}")
def get_note(note_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return note_service.get_note_by_id(db, note_id, user.id)


@router.put("/{note_id}")
def update_note(note_id: str, data: UpdateNoteSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return note_service.update_note(db, note_id, user.id, data.model_dump(exclude_none=True))


@router.delete("/{note_id}")
def delete_note(note_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    note_service.delete_note(db, note_id, user.id)
    return {"message": "Note deleted"}


@router.put("/{note_id}/archive")
def archive_note(note_id: str, data: ArchiveSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return note_service.archive_note(db, note_id, user.id, data.archived)


@router.put("/{note_id}/pin")
def pin_note(note_id: str, data: PinSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return note_service.pin_note(db, note_id, user.id, data.pinned)


@router.post("/{note_id}/share")
def share_note(note_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return note_service.share_note(db, note_id, user.id)


@router.delete("/{note_id}/share")
def unshare_note(note_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    note_service.unshare_note(db, note_id, user.id)
    return {"message": "Note unshared"}
